main/src/ca/uwaterloo/flix/language/phase/CodeGen.scala

```diff
@@ -573,24 +565,69 @@ object CodeGen extends Phase[ExecutableAst.Root, ExecutableAst.Root] {
+       * case is Unit, we compile the case with an object field type to itself and we compile the case with Unit field type
+       * to jvm null.
+       */
+      if(isNullOptimizable(enums, exp.tpe)) {
+        // Compile the expression
+        compileExpression(prefix, functions, declarations, interfaces, enums, visitor, entryPoint)(exp)
+        /*
+         * If type of the expression is Unit, then use the `IFNONNULL` comparator which will jump to `falseLabel` if the
+         * object is not null. If the type of expression is not Unit, then we use `IFNULL` comparator which will jump to
+         * `falseLabel` if the object is null.
+         */
+        val comparator = if(enums(exp.tpe)(tag).fieldType == Type.Unit) {
+          IFNONNULL
+        } else {
+          IFNULL
+        }
+        // Label to the end of the comparison
+        val endLabel = new Label()
+        // Label to `false`
+        val falseLabel = new Label()
+        visitor.visitJumpInsn(comparator, falseLabel)
+        visitor.visitInsn(ICONST_1)
+        visitor.visitJumpInsn(GOTO, endLabel)
+        visitor.visitLabel(falseLabel)
+        visitor.visitInsn(ICONST_0)
+        visitor.visitLabel(endLabel)
+      } else {
+        // First we compile the `exp`
+        compileExpression(prefix, functions, declarations, interfaces, enums, visitor, entryPoint)(exp)
+        // We look in the enum map to find the qualified name of the class of the enum case
+        val qualName = enums(exp.tpe)(tag).qualName
+        // We check if the enum is `instanceof` the class
+        visitor.visitTypeInsn(INSTANCEOF, decorate(qualName))
+      }
 
     case Expression.Tag(enum, tag, exp, tpe, loc) =>
       // Adding source line number for debugging
       addSourceLine(visitor, loc)
-      //  We look in the enum map to find the qualified name of the class of the enum case
-      val (clazzName, enumCase) = enums(tpe, tag)
+      //  We look in the enum map to find the qualified name and executableAst representation of the class of the enum case
+      val clazzName = enums(tpe)(tag).qualName
+      val enumCase = enums(tpe)(tag).executableCase
       /*
-       If the definition of the enum case has a `Unit` field, then it is represented by singleton pattern which means
-       there is only one instance of the class initiated as a field. We have to fetch this field instead of instantiating
-       a new one.
+       * If the enum has two cases, and the field type of one of the cases is an object and the field type of the other
+       * case is Unit, we compile the case with an object field type to itself and we compile the case with Unit field type
+       * to jvm null.
        */
-      if (isSingletonEnum(enumCase)) {
+      if(isNullOptimizable(enums, tpe)) {
+        if(exp.tpe == Type.Unit) {
+          // If the type of the field is Unit, we push a NULL to the top of the stack
+          visitor.visitInsn(ACONST_NULL)
```

Comment:
We must evaluate `exp` since it could have side effects. 

For example, `Option` can be optimized, but the expression:

```
??? : Unit
```

must be evaluated and not just compiled as `null`.